File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.py
# Brute Force Code & Optimal Code
class Solution:
    def deletehelper(self, root, st, result):
        # if the current node is none there is nothing to process
        if root is None:
            return None
        # first process the left subtrees this ensures children are handled before the current node
        root.left = self.deletehelper(root.left, st, result)
        # process the right subtree
        root.right = self.deletehelper(root.right, st, result)
        # checke if the current node needs to be deleted
        if root.val in st:
            # if the left child exists, if becomes the root of new seprated trees
            if root.left is not None:
                result.append(root.left)
            # if right child exists, if becomes the root of new seprated trees
            if root.right is not None:
                result.append(root.right)
        else:
            # current node does not need to be delected so keep it in the tree
            return root
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        # list will store the root of all remaining trees
        result = []
        # convert to_delete into set so checking whether node should be deleted 
        st = set(to_delete)
        # recurively delete all required nodes
        self.deletehelper(root, st, result)
        # deletehelper() only adds children of deleted nodes to result therefore if the original root ifself is not deleted, need to add it manually
        if root.val not in st:
            result.append(root)
        # return the root of all remaining trees
        return result 

# Time Complexity : O(N)
# Space Complexity : O(N)

File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    TreeNode* deletehelper(TreeNode* root, unordered_set<int>& st, vector<TreeNode*>& result) {
        // if the current node is NULL, there is nothing to process
        if (root == NULL) {
            return NULL;
        }
        // first process the left subtree this ensures children are handled before the current node
        root->left = deletehelper(root->left, st, result);
        // process the right subtree
        root->right = deletehelper(root->right, st, result);
        // check if the current node needs to be deleted
        if (st.find(root->val) != st.end()) {
            // if the left child exists it becomes the root of a new separated tree
            if (root->left != NULL) {
                result.push_back(root->left);
            }
            // if the right child exists it becomes the root of a new separated tree
            if (root->right != NULL) {
                result.push_back(root->right);
            }
            // current node is deleted
            return NULL;
        }
        else {
            // current node does not need to be deleted so keep it in the tree
            return root;
        }
    }
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        // list will store the root of all remaining trees
        vector<TreeNode*> result;
        // convert to_delete into a set so checking whether a node should be deleted is O(1) average
        unordered_set<int> st(to_delete.begin(), to_delete.end());
        // recursively delete all required nodes
        deletehelper(root, st, result);
        // deletehelper() only adds children of deleted nodes to result. Therefore, if the original root itself is not deleted, add it manually.
        if (root != NULL && st.find(root->val) == st.end()) {
            result.push_back(root);
        }
        // return the root of all remaining trees
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)